### ISSUE TITLE
ZODB Not Unlocking Properly on Process Fail

### ISSUE THREAD

**Issue by rr326** [zopefoundation/ZODB#166]
This seems similar to [this older issue](https://groups.google.com/forum/#!topic/zodb/UgWIGri4tEY).

I have a process that uses ZODB. It might die unexpectedly. When that happens, sometimes the ZODB lock is properly unlocked when the process dies, and sometimes it is not. 

I only have 1 process at a time using the ZODB, so I'm tempted to just delete the lockfile if I get this error on startup, but that's inelegant. 

Here is a rather elaborate test program [gist](https://gist.github.com/rr326/86b29f496c827964c32d38c97bad1775) that helps play with the error.

OSX 10.12.5
Python 3.5.2

Example error:
```
********************************************************************************
Testing OOBTree
Error locking file /var/folders/8g/j1mly5696m5dhfln7w08kzjm0000gn/T/tmpjve40hmi/zodb_OOBTree.lock; pid=16532
Traceback (most recent call last):
  File "/Users/rrosen/dev/filing/app/python/pyvenv352/lib/python3.5/site-packages/zc/lockfile/__init__.py", line 57, in _lock_file
    fcntl.flock(file.fileno(), _flags)
BlockingIOError: [Errno 35] Resource temporarily unavailable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/rrosen/dev/filing/app/python/pyvenv352/lib/python3.5/site-packages/zc/lockfile/__init__.py", line 88, in __init__
    _lock_file(fp)
  File "/Users/rrosen/dev/filing/app/python/pyvenv352/lib/python3.5/site-packages/zc/lockfile/__init__.py", line 59, in _lock_file
    raise LockError("Couldn't lock %r" % file.name)
zc.lockfile.LockError: Couldn't lock '/var/folders/8g/j1mly5696m5dhfln7w08kzjm0000gn/T/tmpjve40hmi/zodb_OOBTree.lock'
```

Results after 10 trials: ("true" = success, "false" = fail due to lock problem)
```
{
    "OOBTree": {
        "false": 3,
        "true": 7
    },
    "PersistentMapping": {
        "false": 4,
        "true": 6
    }
}

```

Ideally, I'd love to know how to elegantly and properly fix the problem.
Barring that, I'd be happy to know a workaround and if for some reason it is NOT a good idea to just delete the lockfile should I get into this problem. 

Thanks.

**Comment by jimfulton** [comment]
This is almost always a case of the process not being truly dead.  fcntl-based locking  is used and so the lock really should be freed when the process exits.   I suspect there's something going on you're not aware of, but of course, I could be wrong.  Your gist uses multiprocessing. I wonder if it adds any interest to the issue. <shrug>

If you want a quick workaround, and you know there isn't contention, you can rename the file.  Removing the file might not work because the inode may be reused while still locked. (fcntl locking locks inodes.)

What is your platform?

**Comment by rr326** [comment]
My platform is OSX (if that's what you're asking).

Also, this was a simplified version of my real app, which is an Electron app that launches a python backend. When Electron shuts down, the ZODB wouldn't unlock.  

I'll check to see if the process is still running even though it should be dead.  Thanks for the tips.

**Comment by rr326** [comment]
FYI - as far as I can tall the process is actually truly dead.

```
Error locking file /var/folders/8g/j1mly5696m5dhfln7w08kzjm0000gn/T/tmp8aak6yhm/zodb_PersistentMapping.lock; pid=30598

$ ps -p 30598
  PID TTY           TIME CMD
$
```

**Comment by rr326** [comment]
My bad.

Your assumption was correct (This is almost always a case of the process not being truly dead.)

There are a few weird things going on in my code. In my above test code, adding a `sleep(1)` after terminating the process gives the os time to finish killing the process and releasing the lock.

When I figured that out, I was able to find some weird stuff in my code where the process wasn't getting closed as I thought it was. 

Thanks for the clue.
